### PR TITLE
(MODULES-4194) update CRL lastUpdate and nextUpdate fields

### PR DESCRIPTION
**Pull request by adrienthebo**
/cc @nfagerlund this will need docs updates; when the CA cert is rotated the CRL should be updated as well to make sure that it doesn't expire. I'm still working on how to distribute this file.

### CODE REVIEW

**Comment by nfagerlund** [comment]
OK! Holler again once we have a process for doing that.

File: lib/puppet/face/certregen.rb [hunk 1]
```diff
@@ -47,6 +48,15 @@
     end
   end
 
+  action(:crl) do
+    summary 'Update the lastUpdate and nextUpdate field for the CA CRL'
```
Comment (binford2k) [inline lib/puppet/face/certregen.rb]: Could this be a more user friendly description? As an end user, I may not know what those terms mean or what the effect of this change would be.

File: lib/puppet/functions/classified_with.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Puppet::Functions.create_function(:classified_with) do
```
Comment (binford2k) [inline lib/puppet/functions/classified_with.rb]: Should these functions be named `is_classified_with`?

Are these parse-order safe?

File: lib/puppet_x/certregen/crl.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        Puppet::SSL::CertificateRevocationList.indirection.save(crl)
+        FileUtils.cp(Puppet[:cacrl], Puppet[:hostcrl])
+      end
+
+      # @api private
+      def update_valid_time_range_to_start_at(crl_content, time)
+        # The CRL is not valid if the time of checking == the time of last_update.
+        # So to have it valid right now we need to say that it was updated one second ago.
+        crl_content.last_update = time - 1
+        crl_content.next_update = time + FIVE_YEARS
```
Comment (binford2k) [inline lib/puppet_x/certregen/crl.rb]: Should this respect `ca_ttl`?

**Comment by nfagerlund** [comment]
I've got a more existential question. Is there a reason you'd ever want to update the CA and NOT update the CRL? Or update the CRL without updating the CA?

File: lib/puppet/face/certregen.rb (hunk 1 above)
Comment (adrienthebo) [inline lib/puppet/face/certregen.rb]: What wording would you prefer?

File: lib/puppet/functions/classified_with.rb (hunk 2 above)
Comment (adrienthebo) [inline lib/puppet/functions/classified_with.rb]: Sadly there's no yak shaving emote. That being said I'll rename them.

These functions get around the lack of parse order safety in Puppet by querying the classification state of the compiler before compilation even starts. This relies on the fact that masters are almost always classified with `puppet_enterprise::profile::master` via the classifier, so combined with the `defined()` check this is about as good as we'll get without linking the compiler against libtachyon.

**Comment by adrienthebo** [comment]
@nfagerlund fair point. I might wrap up both actions in a `puppet certregen refresh` command. Thoughts?

File: lib/puppet/face/certregen.rb [hunk 4]
```diff
@@ -13,6 +13,25 @@
     expiring CA certificates, and remediation for expired CA certificates.
   EOT
 
+  action(:refresh) do
```
Comment (adrienthebo) [inline lib/puppet/face/certregen.rb]: /cc @nfagerlund

File: lib/puppet_x/certregen/crl.rb (hunk 3 above)
Comment (adrienthebo) [inline lib/puppet_x/certregen/crl.rb]: Good catch. For that it's worth `ca_ttl` is a misnomer as it's the ttl for all signed certificates, so let's not overload it more.

**Comment by binford2k** [comment]
👍

**Comment by camlow325** [comment]
I'm good with the code on this from a first pass.  I'll do a little manual functional testing of it and comment back here if I find anything from that.

**Comment by adrienthebo** [comment]
Changed `manage_crl` to default to true, and I've added a spec for confirming the crl number was incremented.

**Comment by nfagerlund** [comment]
@adrienthebo So, one misunderstanding we've been running into within the company is that people think this module is meant to regenerate ALL certificates in your deployment, cf. the massive PE documents on how to do that. I still don't quite know if that's one of our ambitions for it? 

Putting the CA-specific rotation stuff in a generically-named "refresh" action seems likely to expand that confusion. If the name of the command was "ca_rotate" or something, I'd support that, but given the existing name, I think the action name needs to have something to do with "ca".

**Comment by adrienthebo** [comment]
@nfagerlund I do like the idea of exposing the separate "regenerate CA cert" and "update CRL expiry" as separate commands, but that being said we could keep `puppet certregen ca` as the command to invoke and rename the "regenerate CA cert" command. Does that work for you?

**Comment by nfagerlund** [comment]
@adrienthebo Yeah, I think that's sensible.

**Comment by adrienthebo** [comment]
Pull request updated with a `cacert` action to only update the CA cert, and the `ca` action has been changed to invoke the `cacert` and `crl` actions.

**Comment by adrienthebo** [comment]
And if `puppet certregen ca` is too confusing we can change it, @nfagerlund would you like to change that to `rotate_ca` or somesuch?

/cc @jdwelch opinions on the above?

**Comment by adrienthebo** [comment]
Switched to use the multi-arg form of `file()` in client.pp to determine which CA/CRL file source to use, so we no longer check `$settings::ca`.

**Comment by jdwelch** [comment]
Changes are sensible to me.

**Comment by nfagerlund** [comment]
@adrienthebo I think "ca" is a nice serviceable name for the combined action. 👍

**Comment by adrienthebo** [comment]
@camlow325 I think we're good on the docs and ux front; are there any other changes you'd like to see on this PR before we merge it?

**Pull request merged**